project-koku/koku
# COST-3540: Support new distribution options in cost model

## Jira Ticket

[COST-3540](https://issues.redhat.com/browse/COST-3540)

## Description

This change follows this [pr](https://github.com/project-koku/koku/pull/4178) to validate the `distribution_info` json data or create a default if does not exist.

## Testing

- make sure to have changes in [this pr](https://github.com/project-koku/koku/pull/4178) and run migrations

1. Checkout this branch
2. Restart Koku
3. load data with `make create-test-customer` and `make load-test-customer-data`
4. Once the data has loaded, hit this endpoint, http://127.0.0.1:8000/api/cost-management/v1/cost-models/

    <details>
      <summary>
        You should see the `distribution_info` object for all the cost models returned, similar to this:
      </summary>
      
      ```
        {
            "uuid": "b24f78ce-82f6-444a-8acd-9fb24921dabb",
            "name": "Cost Management AWS Cost Model",
            "description": "A cost model for markup on AWS costs.",
            "source_type": "Amazon Web Services",
            "created_timestamp": "2023-03-06T21:56:08.048869Z",
            "updated_timestamp": "2023-03-06T21:56:08.048889Z",
            "rates": [],
            "markup": {
                "value": "10.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "distribution_info": {
                "distribution_type": "cpu",
                "platform_cost": true,
                "worker_cost": true
            },
            "currency": "USD",
            "source_uuids": [],
            "sources": [
                {
                    "uuid": "9b2cf13e-fe21-4034-9add-5625e45d6cbc",
                    "name": "Test AWS Source"
                }
            ]
        },
        {
            "uuid": "e0632b16-e0cc-45ba-9e51-f6e7cd0af67f",
            "name": "Cost Management Azure Cost Model",
            "description": "A cost model for markup on Azure costs.",
            "source_type": "Microsoft Azure",
            "created_timestamp": "2023-03-06T21:56:18.931336Z",
            "updated_timestamp": "2023-03-06T21:56:18.931410Z",
            "rates": [],
            "markup": {
                "value": "20.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "distribution_info": {
                "distribution_type": "cpu",
                "platform_cost": true,
                "worker_cost": true
            },
            "currency": "USD",
            "source_uuids": [],
            "sources": [
                {
                    "uuid": "46287741-3a11-46a5-815b-ddecac816e29",
                    "name": "Test Azure Source"
                }
            ]
        },
        {
            "uuid": "9633b93e-3b2b-441d-834c-c00c7540914c",
            "name": "Cost Management GCP Cost Model",
            "description": "A cost model for markup on GCP costs.",
            "source_type": "Google Cloud Platform",
            "created_timestamp": "2023-03-06T21:56:30.410878Z",
            "updated_timestamp": "2023-03-06T21:56:30.410904Z",
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "distribution_info": {
                "distribution_type": "cpu",
                "platform_cost": true,
                "worker_cost": true
            },
            "currency": "USD",
            "source_uuids": [],
            "sources": [
                {
                    "uuid": "5063aab6-50b8-4c08-aa3c-8f80b55f8e8e",
                    "name": "Test GCP Source"
                }
            ]
        }
    ``` 
  </details>


## Notes

...
